Expensify/App
# Update JS on desktop and web

Apologies, but I'm switching over from https://github.com/Expensify/ReactNativeChat/pull/274/ to this PR. It's minimal and has a clean commit history that should be easier to follow. Most importantly: it actually works!

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/141663

### Details
This PR introduces a JS utility which periodically checks for updates executes platform-specific update code. It also introduces a `download` network utility, and a Webpack plugin that writes the version number from `package.json` to a file hosted at the site root (to facilitate the update logic for web).

It also cleans up/refactors the build process in a number of ways:

- Unifies the app entry point on all platforms into `/index.js`. Before, the webpack entry point for both web and desktop was `web/index.js`, which was confusing. Platform-specific configurations that need to be run on app launch are consolidated in `src/setup`, allowing us to execute platform-specific code in `index.js`. 
- Moves webpack files into a `webpack` subdirectory
- Moves the Electron main process file `main.js` to `desktop/main.js`, so that it's clear that that file is specific to the desktop platforms.

Note: iOS and Android will likely use a separate library and process to update their JS (via CodePusher), so they aren't considered in this PR as of now.

### Testing steps
1. Run the following commands, and make sure that the apps launch and nothing blows up:

- `npm run ios`
- `npm run android`
- `npm run web`
    - Also, when you do this, go to `http://localhost:8080/version.json`, and verify that the JSON is well-formed and that the version number matches the version from `package.json`
- `npm run desktop`

2. Let's also make sure that the desktop build succeeds.
    - Delete the Chat app
    - Use the diff from [this PR](https://github.com/Expensify/ReactNativeChat/pull/631)'s description to make a local production build work
    - Run `npm run desktop-build`
    - Open the project in finder, double-click on `dist/Chat.dmg`, and install the app. Run it and make sure it works.
    - The logs for the desktop production app live in `~/Library/Logs/react-native-chat/main.log`. Inspect them by any means you prefer and search for the log line "Checking for update". This indicates that the auto-updater is running. Because this is a local production build, you'll likely see a 404 error against a query to Github's servers immediately after. That's expected for a local production build.

3. That's about all we can do to test this for now, because the only way to actually test it is to leave the app running and deploy some new code triggering an actual code update.
